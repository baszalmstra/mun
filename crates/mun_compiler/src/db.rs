use crate::Config;
use mun_codegen::IrDatabase;
use mun_hir::{salsa, HirDatabase};
use std::sync::Arc;

/// A compiler database is a salsa database that enables increment compilation.
#[salsa::database(
    mun_hir::SourceDatabaseStorage,
    mun_hir::DefDatabaseStorage,
    mun_hir::HirDatabaseStorage,
    mun_codegen::IrDatabaseStorage
)]
#[derive(Debug)]
pub struct CompilerDatabase {
    runtime: salsa::Runtime<CompilerDatabase>,
}

impl CompilerDatabase {
    /// Constructs a new database
    pub fn new(config: &Config) -> Self {
        let mut db = CompilerDatabase {
            runtime: salsa::Runtime::default(),
        };

        // Set the initial configuration
        db.set_context(Arc::new(mun_codegen::Context::create()));
        db.set_config(config);

        db
    }

    /// Applies the given configuration to the database
    pub fn set_config(&mut self, config: &Config) {
        self.set_target(config.target.clone());
        self.set_optimization_lvl(config.optimization_lvl);
    }
}

impl salsa::Database for CompilerDatabase {
    fn salsa_runtime(&self) -> &salsa::Runtime<CompilerDatabase> {
        &self.runtime
    }
    fn salsa_runtime_mut(&mut self) -> &mut salsa::Runtime<CompilerDatabase> {
        &mut self.runtime
    }
}
