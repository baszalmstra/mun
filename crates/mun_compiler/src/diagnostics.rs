use crate::diagnostics_snippets::{emit_hir_diagnostic, emit_syntax_error};
use crate::DisplayColor;
use hir::{DiagnosticSink, HirDatabase};
use std::io::Cursor;

/// Emits all diagnostic messages currently in the database; returns true if errors were
/// emitted.
pub fn emit_diagnostics(
    db: &dyn HirDatabase,
    writer: &mut dyn std::io::Write,
    display_color: DisplayColor,
) -> Result<bool, anyhow::Error> {
    let emit_colors = display_color.should_enable();
    let mut has_error = false;

    for package in hir::Package::all(db) {
        for module in package.modules(db) {
            if let Some(file_id) = module.file_id(db) {
                let parse = db.parse(file_id);
                let source_code = db.file_text(file_id);
                let relative_file_path = db.file_relative_path(file_id);
                let line_index = db.line_index(file_id);

                // Emit all syntax diagnostics
                for syntax_error in parse.errors().iter() {
                    emit_syntax_error(
                        syntax_error,
                        relative_file_path.as_str(),
                        &source_code,
                        &line_index,
                        emit_colors,
                        writer,
                    )?;
                    has_error = true;
                }

                // Emit all HIR diagnostics
                let mut error = None;
                module.diagnostics(
                    db,
                    &mut DiagnosticSink::new(|d| {
                        has_error = true;
                        if let Err(e) = emit_hir_diagnostic(d, db, file_id, emit_colors, writer) {
                            error = Some(e)
                        };
                    }),
                );

                // If an error occurred when emitting HIR diagnostics, return early with the error.
                if let Some(e) = error {
                    return Err(e.into());
                }
            }
        }
    }

    Ok(has_error)
}

/// Returns all diagnostics as a human readable string
pub fn emit_diagnostics_to_string(
    db: &dyn HirDatabase,
    display_color: DisplayColor,
) -> anyhow::Result<Option<String>> {
    let mut compiler_errors: Vec<u8> = Vec::new();
    if !emit_diagnostics(db, &mut Cursor::new(&mut compiler_errors), display_color)? {
        Ok(None)
    } else {
        Ok(Some(String::from_utf8(compiler_errors).map_err(|e| {
            anyhow::anyhow!(
                "could not convert compiler diagnostics to valid UTF8: {}",
                e
            )
        })?))
    }
}

#[cfg(test)]
mod tests {
    use super::emit_diagnostics_to_string;
    use crate::{Config, DisplayColor, Driver, PathOrInline, RelativePathBuf};

    /// Compile passed source code and return all compilation errors
    fn compilation_errors(source_code: &str) -> String {
        let config = Config::default();

        let input = PathOrInline::Inline {
            rel_path: RelativePathBuf::from("main.mun"),
            contents: source_code.to_owned(),
        };

        let (driver, _) = Driver::with_file(config, input).unwrap();
        emit_diagnostics_to_string(driver.database(), DisplayColor::Disable)
            .unwrap()
            .unwrap_or_default()
    }

    #[test]
    fn test_syntax_error() {
        insta::assert_display_snapshot!(compilation_errors("\n\nfn main(\n struct Foo\n"));
    }

    #[test]
    fn test_unresolved_value_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn main() {\nlet b = a;\n\nlet d = c;\n}"
        ));
    }

    #[test]
    fn test_unresolved_type_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn main() {\nlet a = Foo{};\n\nlet b = Bar{};\n}"
        ));
    }

    #[test]
    fn test_expected_function_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn main() {\nlet a = Foo();\n\nlet b = Bar();\n}"
        ));
    }

    #[test]
    fn test_mismatched_type_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn main() {\nlet a: f64 = false;\n\nlet b: bool = 22;\n}"
        ));
    }

    #[test]
    fn test_duplicate_definition_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn foo(){}\n\nfn foo(){}\n\nstruct Bar;\n\nstruct Bar;\n\nfn BAZ(){}\n\nstruct BAZ;"
        ));
    }

    #[test]
    fn test_possibly_uninitialized_variable_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn main() {\nlet a;\nif 5>6 {\na = 5\n}\nlet b = a;\n}"
        ));
    }

    #[test]
    fn test_access_unknown_field_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nstruct Foo {\ni: bool\n}\n\nfn main() {\nlet a = Foo { i: false };\nlet b = a.t;\n}"
        ));
    }

    #[test]
    fn test_free_type_alias_error() {
        insta::assert_display_snapshot!(compilation_errors("\n\ntype Foo;"));
    }

    #[test]
    fn test_type_alias_target_undeclared_error() {
        insta::assert_display_snapshot!(compilation_errors("\n\ntype Foo = UnknownType;"));
    }

    #[test]
    fn test_cyclic_type_alias_error() {
        insta::assert_display_snapshot!(compilation_errors("\n\ntype Foo = Foo;"));
    }

    #[test]
    fn test_expected_function() {
        insta::assert_display_snapshot!(compilation_errors("\n\nfn foo() { let a = 3; a(); }"));
    }
}
