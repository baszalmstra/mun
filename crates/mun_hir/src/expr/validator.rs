use crate::code_model::src::HasSource;
use crate::diagnostics::{
    ExportedPrivate, ExternCannotHaveBody, ExternNonPrimitiveParam, FreeTypeAliasWithoutTypeRef,
};
use crate::expr::BodySourceMap;
use crate::in_file::InFile;
use crate::resolve::HasResolver;
use crate::{
    diagnostics::DiagnosticSink, Body, Expr, Function, HirDatabase, InferenceResult, TypeAlias,
};
use crate::{HasVisibility, Ty, Visibility};

use mun_syntax::{AstNode, SyntaxNodePtr};
use std::sync::Arc;

mod literal_out_of_range;
mod uninitialized_access;

#[cfg(test)]
mod tests;

pub struct ExprValidator<'a> {
    func: Function,
    infer: Arc<InferenceResult>,
    body: Arc<Body>,
    body_source_map: Arc<BodySourceMap>,
    db: &'a dyn HirDatabase,
}

impl<'a> ExprValidator<'a> {
    pub fn new(func: Function, db: &'a dyn HirDatabase) -> Self {
        let (body, body_source_map) = db.body_with_source_map(func.id.into());
        ExprValidator {
            func,
            db,
            infer: db.infer(func.id.into()),
            body,
            body_source_map,
        }
    }

    pub fn validate_body(&self, sink: &mut DiagnosticSink) {
        self.validate_literal_ranges(sink);
        self.validate_uninitialized_access(sink);
        self.validate_extern(sink);
        self.validate_privacy(sink);
    }

    pub fn validate_privacy(&self, sink: &mut DiagnosticSink) {
        let resolver = self.func.id.resolver(self.db.upcast());
        let fn_data = self.func.data(self.db.upcast());
        let ret_type_ref = fn_data.ret_type();
        let param_types = fn_data
            .params()
            .iter()
            .chain(std::iter::once(ret_type_ref))
            .map(|type_ref| {
                let (ty, _) = Ty::from_hir(self.db, &resolver, fn_data.type_ref_map(), *type_ref);
                (ty, type_ref)
            });

        let fn_visibility = self.func.visibility(self.db);
        let type_is_allowed = |ty: &Ty| match fn_visibility {
            Visibility::Module(module_id) => {
                ty.visibility(self.db).is_visible_from(self.db, module_id)
            }
            Visibility::Public => ty.visibility(self.db).is_externally_visible(),
        };

        let file_id = self.func.source(self.db.upcast()).file_id;
        param_types
            .filter(|(ty, _)| !type_is_allowed(ty))
            .for_each(|(_, type_ref)| {
                sink.push(ExportedPrivate {
                    file: file_id,
                    type_ref: fn_data
                        .type_ref_source_map()
                        .type_ref_syntax(*type_ref)
                        .unwrap(),
                })
            });
    }

    pub fn validate_extern(&self, sink: &mut DiagnosticSink) {
        if !self.func.is_extern(self.db) {
            return;
        }

        // Validate that there is no body
        match self.body[self.func.body(self.db).body_expr] {
            Expr::Missing => {}
            _ => sink.push(ExternCannotHaveBody {
                func: self
                    .func
                    .source(self.db.upcast())
                    .map(|f| SyntaxNodePtr::new(f.syntax())),
            }),
        }

        if let Some(sig) = self.func.ty(self.db).callable_sig(self.db) {
            let fn_data = self.func.data(self.db.upcast());
            for (arg_ty, ty_ref) in sig.params().iter().zip(fn_data.params()) {
                if arg_ty.as_struct().is_some() {
                    let arg_ptr = fn_data
                        .type_ref_source_map()
                        .type_ref_syntax(*ty_ref)
                        .map(|ptr| ptr.syntax_node_ptr())
                        .unwrap();
                    sink.push(ExternNonPrimitiveParam {
                        param: InFile::new(self.func.source(self.db.upcast()).file_id, arg_ptr),
                    })
                }
            }

            let return_ty = sig.ret();
            if return_ty.as_struct().is_some() {
                let arg_ptr = fn_data
                    .type_ref_source_map()
                    .type_ref_syntax(*fn_data.ret_type())
                    .map(|ptr| ptr.syntax_node_ptr())
                    .unwrap();
                sink.push(ExternNonPrimitiveParam {
                    param: InFile::new(self.func.source(self.db.upcast()).file_id, arg_ptr),
                })
            }
        }
    }
}

pub struct TypeAliasValidator<'a> {
    type_alias: TypeAlias,
    db: &'a dyn HirDatabase,
}

impl<'a> TypeAliasValidator<'a> {
    /// Constructs a validator for the provided `TypeAlias`.
    pub fn new(type_alias: TypeAlias, db: &'a dyn HirDatabase) -> Self {
        TypeAliasValidator { type_alias, db }
    }

    /// Validates that the provided `TypeAlias` has a target type of alias.
    pub fn validate_target_type_existence(&self, sink: &mut DiagnosticSink) {
        let src = self.type_alias.source(self.db.upcast());
        if src.value.type_ref().is_none() {
            sink.push(FreeTypeAliasWithoutTypeRef {
                type_alias_def: src.map(|t| SyntaxNodePtr::new(t.syntax())),
            })
        }
    }
}
