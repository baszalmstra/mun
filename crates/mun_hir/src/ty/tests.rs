use crate::db::SourceDatabase;
use crate::diagnostics::DiagnosticSink;
use crate::expr::BodySourceMap;
use crate::ids::LocationCtx;
use crate::mock::MockDatabase;
use crate::{Function, HirDisplay, InferenceResult};
use mun_syntax::{ast, AstNode};
use std::fmt::Write;
use std::sync::Arc;

#[test]
fn infer_conditional_return() {
    infer_snapshot(
        r#"
    fn foo(a:int):int {
        if a > 4 {
            return 4;
        }
        a
    }

    fn bar(a:int):int {
        if a > 4 {
            return 4;
        } else {
            return 1;
        }
    }
    "#,
    )
}

#[test]
fn infer_return() {
    infer_snapshot(
        r#"
    fn test():int {
        return; // error: mismatched type
        return 5;
    }
    "#,
    )
}

#[test]
fn infer_basics() {
    infer_snapshot(
        r#"
    fn test(a:int, b:float, c:never, d:bool): bool {
        a;
        b;
        c;
        d
    }
    "#,
    )
}

#[test]
fn infer_branching() {
    infer_snapshot(
        r#"
    fn test() {
        let a = if true { 3 } else { 4 }
        let b = if true { 3 }               // Missing else branch
        let c = if true { 3; }
        let d = if true { 5 } else if false { 3 } else { 4 }
        let e = if true { 5.0 } else { 5 }  // Mismatched branches
    }
    "#,
    )
}

#[test]
fn void_return() {
    infer_snapshot(
        r#"
    fn bar() {
        let a = 3;
    }
    fn foo(a:int) {
        let c = bar()
    }
    "#,
    )
}

#[test]
fn place_expressions() {
    infer_snapshot(
        r#"
    fn foo(a:int) {
        a += 3;
        3 = 5; // error: invalid left hand side of expression
    }
    "#,
    )
}

#[test]
fn update_operators() {
    infer_snapshot(
        r#"
    fn foo(a:int, b:float) {
        a += 3;
        a -= 3;
        a *= 3;
        a /= 3;
        b += 3.0;
        b -= 3.0;
        b *= 3.0;
        b /= 3.0;
        a *= 3.0; // mismatched type
        b *= 3; // mismatched type
    }
    "#,
    )
}

#[test]
fn infer_loop() {
    infer_snapshot(
        r#"
    fn foo() {
        loop {}
    }
    "#,
    )
}

#[test]
fn infer_break() {
    infer_snapshot(
        r#"
    fn foo():int {
        break; // error: not in a loop
        loop { break 3; break 3.0; } // error: mismatched type
        let a:int = loop { break 3.0; } // error: mismatched type
        loop { break 3; }
        let a:int = loop { break loop { break 3; } }
        loop { break loop { break 3.0; } } // error: mismatched type
    }
    "#,
    )
}

#[test]
fn infer_while() {
    infer_snapshot(
        r#"
    fn foo() {
        let n = 0;
        while n < 3 { n += 1; };
        while n < 3 { n += 1; break; };
        while n < 3 { break 3; };   // error: break with value can only appear in a loop
        while n < 3 { loop { break 3; }; };
    }
    "#,
    )
}

#[test]
fn invalid_binary_ops() {
    infer_snapshot(
        r#"
    fn foo() {
        let b = false;
        let n = 1;
        let _ = b + n; // error: invalid binary operation
    }
    "#,
    )
}

#[test]
fn struct_decl() {
    infer_snapshot(
        r#"
    struct Foo;
    struct(gc) Bar {
        f: float,
        i: int,
    }
    struct(value) Baz(float, int);


    fn main() {
        let foo: Foo;
        let bar: Bar;
        let baz: Baz;
    }
    "#,
    )
}

#[test]
fn struct_lit() {
    infer_snapshot(
        r#"
    struct Foo;
    struct Bar {
        a: float,
    }
    struct Baz(float, int);

    fn main() {
        let a: Foo = Foo;
        let b: Bar = Bar { a: 1.23, };
        let c = Baz(1.23, 1);

        let a = Foo{}; // error: mismatched struct literal kind. expected `unit struct`, found `record`
        let a = Foo(); // error: mismatched struct literal kind. expected `unit struct`, found `tuple`
        let b = Bar; // error: mismatched struct literal kind. expected `record`, found `unit struct`
        let b = Bar(); // error: mismatched struct literal kind. expected `record`, found `tuple`
        let b = Bar{}; // error: missing record fields: a
        let c = Baz; // error: mismatched struct literal kind. expected `tuple`, found `unit struct`
        let c = Baz{}; // error: mismatched struct literal kind. expected `tuple`, found `record`
        let c = Baz(); // error: this tuple struct literal has 2 fields but 0 fields were supplied
    }
    "#,
    )
}

#[test]
fn struct_field_index() {
    infer_snapshot(
        r#"
    struct Foo {
        a: float,
        b: int,
    }
    struct Bar(float, int)
    struct Baz;

    fn main() {
        let foo = Foo { a: 1.23, b: 4 };
        foo.a
        foo.b
        foo.c // error: attempted to access a non-existent field in a struct.
        let bar = Bar(1.23, 4);
        bar.0
        bar.1
        bar.2 // error: attempted to access a non-existent field in a struct.
        let baz = Baz;
        baz.a // error: attempted to access a non-existent field in a struct.
        let f = 1.0
        f.0; // error: attempted to access a field on a primitive type.
    }
    "#,
    )
}

#[test]
fn primitives() {
    infer_snapshot(
        r#"
    fn unsigned_primitives(a: u8, b: u16, c: u32, d: u64, e: usize, f: uint) : u8 { a }
    fn signed_primitives(a: i8, b: i16, c: i32, d: i64, e: isize, f: int) : i8 { a }
    fn float_primitives(a: f32, b: f64, c: float) : f32 { a }
    "#,
    )
}

fn infer_snapshot(text: &str) {
    let text = text.trim().replace("\n    ", "\n");
    insta::assert_snapshot!(insta::_macro_support::AutoName, infer(&text), &text);
}

fn infer(content: &str) -> String {
    let (db, file_id) = MockDatabase::with_single_file(content);
    let source_file = db.parse(file_id).ok().unwrap();

    let mut acc = String::new();

    let mut infer_def = |infer_result: Arc<InferenceResult>,
                         body_source_map: Arc<BodySourceMap>| {
        let mut types = Vec::new();

        for (pat, ty) in infer_result.type_of_pat.iter() {
            let syntax_ptr = match body_source_map.pat_syntax(pat) {
                Some(sp) => sp.map(|ast| ast.syntax_node_ptr()),
                None => continue,
            };
            types.push((syntax_ptr, ty));
        }

        for (expr, ty) in infer_result.type_of_expr.iter() {
            let syntax_ptr = match body_source_map.expr_syntax(expr) {
                Some(sp) => {
                    sp.map(|ast| ast.either(|it| it.syntax_node_ptr(), |it| it.syntax_node_ptr()))
                }
                None => continue,
            };
            types.push((syntax_ptr, ty));
        }

        // Sort ranges for consistency
        types.sort_by_key(|(src_ptr, _)| {
            (src_ptr.value.range().start(), src_ptr.value.range().end())
        });
        for (src_ptr, ty) in &types {
            let node = src_ptr.value.to_node(&src_ptr.file_syntax(&db));

            let (range, text) = (
                src_ptr.value.range(),
                node.text().to_string().replace("\n", " "),
            );
            write!(
                acc,
                "{} '{}': {}\n",
                range,
                ellipsize(text, 15),
                ty.display(&db)
            )
            .unwrap();
        }
    };

    let mut diags = String::new();

    let mut diag_sink = DiagnosticSink::new(|diag| {
        write!(diags, "{}: {}\n", diag.highlight_range(), diag.message()).unwrap();
    });

    let ctx = LocationCtx::new(&db, file_id);
    for node in source_file.syntax().descendants() {
        if let Some(def) = ast::FunctionDef::cast(node.clone()) {
            let fun = Function {
                id: ctx.to_def(&def),
            };
            let source_map = fun.body_source_map(&db);
            let infer_result = fun.infer(&db);

            for diag in infer_result.diagnostics.iter() {
                diag.add_to(&db, fun, &mut diag_sink);
            }

            infer_def(infer_result, source_map);
        }
    }

    drop(diag_sink);

    acc.truncate(acc.trim_end().len());
    diags.truncate(diags.trim_end().len());
    [diags, acc].join("\n").trim().to_string()
}

fn ellipsize(mut text: String, max_len: usize) -> String {
    if text.len() <= max_len {
        return text;
    }
    let ellipsis = "...";
    let e_len = ellipsis.len();
    let mut prefix_len = (max_len - e_len) / 2;
    while !text.is_char_boundary(prefix_len) {
        prefix_len += 1;
    }
    let mut suffix_len = max_len - e_len - prefix_len;
    while !text.is_char_boundary(text.len() - suffix_len) {
        suffix_len += 1;
    }
    text.replace_range(prefix_len..text.len() - suffix_len, ellipsis);
    text
}
