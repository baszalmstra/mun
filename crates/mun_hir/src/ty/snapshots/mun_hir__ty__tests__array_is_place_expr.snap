---
source: crates/mun_hir/src/ty/tests.rs
expression: "fn main() {\n    let a = [1,2,3,4,5]\n    a = [5,6,7]\n    a[0] = 0;\n    [1,2,3][0] = 4\n}"
---
10..86 '{     ... = 4 }': ()
20..21 'a': [i32]
24..35 '[1,2,3,4,5]': [i32]
25..26 '1': i32
27..28 '2': i32
29..30 '3': i32
31..32 '4': i32
33..34 '5': i32
40..41 'a': [i32]
40..51 'a = [5,6,7]': ()
44..51 '[5,6,7]': [i32]
45..46 '5': i32
47..48 '6': i32
49..50 '7': i32
56..57 'a': [i32]
56..60 'a[0]': i32
56..64 'a[0] = 0': ()
58..59 '0': i32
63..64 '0': i32
70..77 '[1,2,3]': [i32]
70..80 '[1,2,3][0]': i32
70..84 '[1,2,3][0] = 4': ()
71..72 '1': i32
73..74 '2': i32
75..76 '3': i32
78..79 '0': i32
83..84 '4': i32
